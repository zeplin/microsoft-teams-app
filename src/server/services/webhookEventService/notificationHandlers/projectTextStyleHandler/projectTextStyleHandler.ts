import {
    ProjectTextStyleCreatedEvent,
    ProjectTextStyleUpdatedEvent,
    WebhookEvent
} from "@zeplin/sdk";
import { NotificationHandler } from "../NotificationHandler";
import { SHORT_DELAY } from "../constants";
import { commonTeamsCard, MessageCard } from "../teamsCardTemplates";
import { md } from "../md";
import { getRandomEmoji } from "../getRandomEmoji";
import { getRedirectURLForZeplinApp, getWebAppURL } from "../zeplinURL";
import { getTextStyleUpdateMessage } from "../getStyleUpdateMessage";

type Event = ProjectTextStyleCreatedEvent | ProjectTextStyleUpdatedEvent;

class ProjectTextStyleHandler extends NotificationHandler<Event> {
    delay = SHORT_DELAY;
    private getText(events: Event[]): string {
        const [{
            action,
            context: {
                project: {
                    name: projectName
                }
            },
            resource: {
                data: {
                    name: pivotTextStyleName
                }
            }
        }] = events;
        const actionText = action === "created" ? "added to" : "updated in";
        return events.length === 1
            ? md`**${pivotTextStyleName}** is ${actionText} _${projectName}_! ${getRandomEmoji()}`
            : md`**${events.length} text styles** are ${actionText} _${projectName}_! ${getRandomEmoji()}`;
    }

    private getSectionText(events: Event[]): string {
        const [{
            context: {
                project: {
                    platform: projectPlatform
                }
            }
        }] = events;

        return getTextStyleUpdateMessage(projectPlatform);
    }

    private getWebappURL(
        events: Event[]
    ): string {
        const [{
            context: {
                project: {
                    id: projectId
                }
            }
        }] = events;
        const pathname = `project/${projectId}/styleguide/textstyles`;
        const searchParams = {
            tsid: events.map(({ resource: { id } }) => id)
        };

        return getWebAppURL(pathname, searchParams);
    }

    private getZeplinAppURI(
        events: Event[]
    ): string {
        const [{
            context: {
                project: {
                    id: projectId
                }
            }
        }] = events;
        const searchParams = {
            pid: projectId,
            tsids: events.map(({ resource: { id } }) => id)
        };

        return getRedirectURLForZeplinApp("textStyles", searchParams);
    }

    shouldHandleEvent(event: WebhookEvent): event is Event {
        return event.action !== "deleted";
    }

    getTeamsMessage(
        events: Event[]
    ): MessageCard {
        return commonTeamsCard({
            text: this.getText(events),
            section: {
                text: this.getSectionText(events)
            },
            links: [{
                title: "Open in App",
                url: this.getZeplinAppURI(events)
            }, {
                title: "Open in Web",
                url: this.getWebappURL(events)
            }]
        });
    }
}

export const projectTextStyleHandler = new ProjectTextStyleHandler();
