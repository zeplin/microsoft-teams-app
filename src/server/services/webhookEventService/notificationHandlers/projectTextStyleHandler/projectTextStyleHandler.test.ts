import {
    Project,
    ProjectTextStyleCreatedEvent,
    ProjectTextStyleUpdatedEvent
} from "@zeplin/sdk";

import { projectTextStyleHandler } from "./projectTextStyleHandler";

type GetDummyEventParams = {
    action?: string;
    textStyleId?: string;
    textStyleName?: string;
    projectPlatform?: Project["platform"];
}

function getDummyEvent({
    action = "created",
    textStyleId = "textStyleId",
    textStyleName = "textStyleName",
    projectPlatform = "web"
}: GetDummyEventParams = {}): ProjectTextStyleCreatedEvent | ProjectTextStyleUpdatedEvent {
    return {
        action,
        context: {
            project: {
                id: "projectId",
                name: "projectName",
                platform: projectPlatform
            }
        },
        resource: {
            id: textStyleId,
            data: {
                id: textStyleId,
                name: textStyleName
            }
        }
    } as ProjectTextStyleCreatedEvent | ProjectTextStyleUpdatedEvent;
}

describe("projectTextStyleHandler", () => {
    describe("getTeamsMessage method", () => {
        beforeAll(() => {
            jest.spyOn(Math, "random").mockReturnValue(0);
        });

        describe("for created notification", () => {
            it("should match snapshot when there is only 1 event", () => {
                expect(
                    projectTextStyleHandler.getTeamsMessage([getDummyEvent()])
                ).toMatchSnapshot();
            });

            it("should match snapshot when there are more than 1 events", () => {
                expect(
                    projectTextStyleHandler.getTeamsMessage([
                        getDummyEvent(),
                        getDummyEvent({
                            textStyleId: "anotherTextStyleId",
                            textStyleName: "anotherTextStyleName"
                        })
                    ])
                ).toMatchSnapshot();
            });

            it.each<Project["platform"]>([
                "android",
                "ios",
                "macos",
                "web"
            ])("should match snapshot when project platform is %s",
                projectPlatform => {
                    expect(
                        projectTextStyleHandler.getTeamsMessage([getDummyEvent({ projectPlatform })])
                    ).toMatchSnapshot();
                }
            );
        });

        describe("for updated notification", () => {
            it("should match snapshot when there is only 1 event", () => {
                expect(
                    projectTextStyleHandler.getTeamsMessage([getDummyEvent({ action: "updated" })])
                ).toMatchSnapshot();
            });

            it("should match snapshot when there are more than 1 events", () => {
                expect(
                    projectTextStyleHandler.getTeamsMessage([
                        getDummyEvent({
                            action: "updated"
                        }),
                        getDummyEvent({
                            action: "updated",
                            textStyleId: "anotherTextStyleId",
                            textStyleName: "anotherTextStyleName"
                        })
                    ])
                ).toMatchSnapshot();
            });

            it.each<Project["platform"]>([
                "android",
                "ios",
                "macos",
                "web"
            ])("should match snapshot when project platform is %s",
                projectPlatform => {
                    expect(
                        projectTextStyleHandler.getTeamsMessage([
                            getDummyEvent({
                                action: "updated",
                                projectPlatform
                            })
                        ])
                    ).toMatchSnapshot();
                }
            );
        });
    });
});
