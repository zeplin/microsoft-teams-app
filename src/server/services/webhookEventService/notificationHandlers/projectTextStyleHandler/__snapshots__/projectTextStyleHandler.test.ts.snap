// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projectTextStyleHandler getTeamsMessage method for created notification should match snapshot for every possible project platform 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**textStyleName** is added to _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure \`fonts.xml\` is up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2FtextStyles%3Fpid%3DprojectId%26tsids%3DtextStyleId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/textstyles?tsid=textStyleId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectTextStyleHandler getTeamsMessage method for created notification should match snapshot for every possible project platform 2`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**textStyleName** is added to _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure \`UIFont\` category / extension is up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2FtextStyles%3Fpid%3DprojectId%26tsids%3DtextStyleId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/textstyles?tsid=textStyleId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectTextStyleHandler getTeamsMessage method for created notification should match snapshot for every possible project platform 3`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**textStyleName** is added to _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure \`NSFont\` category / extension is up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2FtextStyles%3Fpid%3DprojectId%26tsids%3DtextStyleId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/textstyles?tsid=textStyleId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectTextStyleHandler getTeamsMessage method for created notification should match snapshot for every possible project platform 4`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**textStyleName** is added to _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2FtextStyles%3Fpid%3DprojectId%26tsids%3DtextStyleId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/textstyles?tsid=textStyleId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectTextStyleHandler getTeamsMessage method for created notification should match snapshot when there are more than 1 events 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**2 text styles** are added to _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2FtextStyles%3Fpid%3DprojectId%26tsids%3DtextStyleId%2CanotherTextStyleId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/textstyles?tsid=textStyleId&tsid=anotherTextStyleId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectTextStyleHandler getTeamsMessage method for created notification should match snapshot when there is only 1 event 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**textStyleName** is added to _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2FtextStyles%3Fpid%3DprojectId%26tsids%3DtextStyleId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/textstyles?tsid=textStyleId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectTextStyleHandler getTeamsMessage method for updated notification should match snapshot for every possible project platform 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**textStyleName** is updated in _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure \`fonts.xml\` is up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2FtextStyles%3Fpid%3DprojectId%26tsids%3DtextStyleId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/textstyles?tsid=textStyleId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectTextStyleHandler getTeamsMessage method for updated notification should match snapshot for every possible project platform 2`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**textStyleName** is updated in _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure \`UIFont\` category / extension is up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2FtextStyles%3Fpid%3DprojectId%26tsids%3DtextStyleId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/textstyles?tsid=textStyleId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectTextStyleHandler getTeamsMessage method for updated notification should match snapshot for every possible project platform 3`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**textStyleName** is updated in _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure \`NSFont\` category / extension is up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2FtextStyles%3Fpid%3DprojectId%26tsids%3DtextStyleId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/textstyles?tsid=textStyleId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectTextStyleHandler getTeamsMessage method for updated notification should match snapshot for every possible project platform 4`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**textStyleName** is updated in _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2FtextStyles%3Fpid%3DprojectId%26tsids%3DtextStyleId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/textstyles?tsid=textStyleId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectTextStyleHandler getTeamsMessage method for updated notification should match snapshot when there are more than 1 events 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**2 text styles** are updated in _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2FtextStyles%3Fpid%3DprojectId%26tsids%3DtextStyleId%2CanotherTextStyleId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/textstyles?tsid=textStyleId&tsid=anotherTextStyleId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectTextStyleHandler getTeamsMessage method for updated notification should match snapshot when there is only 1 event 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**textStyleName** is updated in _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2FtextStyles%3Fpid%3DprojectId%26tsids%3DtextStyleId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/textstyles?tsid=textStyleId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;
